proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+    The sole responsibility of a code owner is to ensure that a commit to their area of the code is appropriately reviewed, either by themself or by someone else.
+
+
+### Code of Conduct
+
+All participants in the LLVM Community must follow the LLVM Community Code of Conduct. That applies to the representatives of governance bodies just as it does to any other individual in the community. The Code of Conduct Committee can suspend or revoke the privilege of any individual to participate in the LLVM Governance process for violations of the Code of Conduct.
+
+Misconduct in elections or governance roles such as fraudulent voting are violations of the Code of Conduct.
+
+
+### Voting Contributors
+
+The first structure is to define the _voting contributor_ base. The _voting contributor_ base seeks to represent the active project contributors. In the context of project governance, the only responsibility of a _voting contributor_ is to vote to elect members of the _area teams_.
+
+The LLVM Foundation will maintain a voter database that identifies voters by name, email, and the accounts used on services used by the LLVM project. The voter database will only be used for election-related purposes. A full privacy policy that complies with appropriate laws will be published, and it will include a process for removing your personal information from the voter database. To be a _voting contributor_ an individual must register with the LLVM Foundation voter database, and meet eligibility requirements.
+
+An individual may be deemed ineligible to vote for violations of the community Code of Conduct. Acting fraudulently or any other method of subverting an LLVM election is a violation of the Code of Conduct.
+
+Voter eligibility is based on recent community participation in the form of _meaningful contributions_. A _meaningful contribution_ may be a pull request contribution, or it could be other activities that help the community function like screening issues, moderating community spaces, hosting community events, or participating in code reviews.
+
+7 days before voting begins a list of eligible voters is constructed from the registered voters including only those that have participated in LLVM community interactions in the preceding 12 months.
+
+Using the methodologies defined in Appendix 2, and only looking at the main LLVM repository, we have 1667 eligible _voting contributors_ based on code contributions alone. This number should expand as we incorporate participation in code reviews, issues, and discussion forums.
+
+
+#### Alternative Considered: Commit Access
+
+Alternatively, the _voting contributor_ base could be defined by commit access. This methodology has two flaws. One is that our commit access list gets stale, and is unwieldy to maintain. The other problem is that with the move to GitHub PRs, commit access is significantly less important. We may begin getting more contributions from users that never bother to request commit privileges.
```

Comment:
I did not consider that. I think limiting voters by affiliation would be hostile to organizational involvement in LLVM which goes against the core project principle that LLVM be friendly to corporate contributors.

Reply:
This is very similar to your other comment about "balancing corporate votes". If you take away voting privileges from individuals because they work for a corporation you're adopting a policy that is hostile to corporations. Corporations are less likely to participate if they know their voices mean less proportionately.

Given how many large corporations provide _massive_ development efforts to LLVM I think it would be an extreme disservice to the project to adopt a policy that curbed their influence because it would encourage them to fork LLVM or go their own routes instead of participating in the community.

Reply:
I am not a native speaker, please do not pick on my poor choice of the word "corporate" here. I am confused how my comment has been understood. Nobody is suggesting taking away anything from anybody. Quite the opposite ensuring that the decision making body is diversified offering inclusive environment.

Perhaps I can illustrate that with an example. It would probably be easy for a malicious actor to get funding and hire N (sufficiently large number) developers (maybe in a low income country) to develop a new LLVM target, for example. By sheer capacity this entity will dominate the backend and if wished to block development in that area. That example is hypothetical but quite possible in my point of view.

I am trying to outline the need to protect ourselves from such situations to the aid of everybody including the large entities which may not be the largest. There are multiple comments in that vain to diversify by skillset, etc, which perhaps goes in a similar direction.

The rules of the community as proposed here should encourage respecting the diversity of LLVM use cases no matter who pays most today. The next tricky question is how can you evaluate what would be a massive development effort? Is that 10 people put by a private entity for 2 year (that can disappear in the first quarter of year 1) or 2 people put by a national lab for 10 years (that are likely to stay beyond that)? 

Different entities with different funding streams have different pros and cons -- giving most power to the strongest of the day does not sounds like an optimal solution.

Reply:
> Different entities with different funding streams have different pros and cons -- giving most power to the strongest of the day does not sounds like an optimal solution.

I think there is no optimal solution; there is actual research into voting systems to demonstrate that, which I read decades ago. At some point we have to go for good-enough, and if it blows up, we need to try again. I don't think any company would be motivated to take over governance of LLVM, it would be cheaper to fork it and go their own way.

Reply:
> That example is hypothetical but quite possible in my point of view.

We have most back-ends covered today and I don't see this being a problem in any real context.

More generally, this strategy would (or should) fall foul of the code of conduct, in which case we don't need extra policies in place to protect against.

Am I being too naive?

Reply:
I agree with @rengolin that any actively malicious behavior should be controllable via the code of conduct, and the proposal does explicitly state that the CoC applies to all governance activities.

I think anything you do to prevent a company from hiring 100 engineers to gain influence over part of the project would back-fire by diluting the influence of well-meaning companies and deter corporate participation.

Reply:
I do think the underlying concern is a legitimate one, not because of any extreme examples where a company goes out of its way to be specifically malicious. It's great if company X invests a lot into LLVM overall, but if an area team then ends up being dominated by folks from company X, it is fair to say that that is almost certainly a bad outcome -- because they will be biased. (What does "dominated" mean? I'm not sure, but I think if people from X are an absolute majority in an area team, that should raise some eyebrows.)

At its core, this is about ensuring that a diversity of viewpoints is represented and, quite frankly, at least for technically-oriented area teams, what company somebody works for is probably more relevant there on average than any other consideration.

On the other hand, I suspect that attempting to cure this with a formal mechanism is likely worse than the (risk of) disease.

We should be able to agree though that there is a potential bad outcome here and that *as a matter of LLVM culture* we want to avoid it.

Reply:
I think if we want to impose limits on corporate representation, it should be done by limiting representation on the sub-teams/steering committee rather than on the voters.  I know some other projects do this.  I'm personally OK if we have no limits at all on representation, but I think limiting voters is not the right way to go.